jcl/src/java.base/share/classes/java/lang/Thread.java

```diff
@@ -1172,19 +1177,18 @@ public synchronized void start() {
 	 *
 	 * Release the lock before calling threadgroup's remove method for this thread.
 	/*[ENDIF]*/
+	synchronized (lock) {
+		if (started) {
+			/*[MSG "K0341", "Thread is already started"]*/
+			throw new IllegalThreadStateException(com.ibm.oti.util.Msg.getString("K0341")); //$NON-NLS-1$
+		}
+		/*[PR 115667, CMVC 94448] In 1.5, thread is added to ThreadGroup when started */
+		group.add(this);
+	}
 
 	try {
-		synchronized(lock) {
-			if (started) {
-				/*[MSG "K0341", "Thread is already started"]*/
-				throw new IllegalThreadStateException(com.ibm.oti.util.Msg.getString("K0341")); //$NON-NLS-1$
-			}
-
-			/*[PR 115667, CMVC 94448] In 1.5, thread is added to ThreadGroup when started */
-			group.add(this);
-
+		synchronized (lock) {
```

Comment:
The lock should not be dropped here: it will allow this thread to be observed as part of the group even though it has not yet been started.

Reply:
I was referring to visibility from the perspective of the `ThreadGroup`, but it appears that only threads that report `isAlive()` are visible through API there.

Reply:
Note that there's a cost to dropping and reacquiring the lock. Unless you can justify that cost, I still think the synchronized blocks should be merged.

Reply:
> I still think the synchronized blocks should be merged.

To move `group.add(this)` out of the `try` block, and keep one `synchronized (lock)`, I think checking `started` & `group.add(this)` don't need to be synchronized on `lock` unless you see a problem.
The `try/finally` should not be moved into one synchronized block, I believe that is the path to a deadlock.

Reply:
Please explain the sequence of thread actions that can lead to deadlock - I don't see a problem.

Reply:
I first note that the inner `synchronized` block is redundant, so we should be discussing this:
```
    synchronized (lock) {
        if (started) {
            throw new IllegalThreadStateException(com.ibm.oti.util.Msg.getString("K0341")); //$NON-NLS-1$
        }
        group.add(this);
        boolean success = false;
        try {
            startImpl();
            success = true;
        } finally {
            if (!success) {
                group.remove(this);
            }
        }
    }
```
There are 3 types of objects being synchronized:
1. Thread
2. Thread.lock
3. ThreadGroup.childrenThreadsLock

The fundamental difficulty is due to inconsistent priority of lock acquisition:
* `Thread.start()` acquires them in the order 1, 2 then 3
* `ThreadGroup.activeCount()` acquires them in the order 3 then 2

A general rule is that resources must be locked in a consistent order to avoid deadlock. The subtle detail that deserves to be documented is that `start()` avoids deadlock by locking 2 and 3 in the reverse order is because no other thread will try to acquire 2 after locking 3 due to the fact the the thread being started is not yet a child of the group. This is not the case in a call to `ThreadGroup.remove()` so the lock on 2 must not be held when that call is made.

This fragility could be avoided by having `Thread.start()` acquire the locks in a consistent order: 1, 3 then 2. To do that, `Thread` would need access to the `childrenThreadsLock` object.

Reply:
I think `Thread.start()` can put `group.remove()` out of `synchronized (lock)` and avoid nesting synchronization. `synchronized (lock)` is not required by `started` &  `group.add()` access, or like existing code, when `group.add()` throws an exception which probably is in slow path, invoke `group.remove()` anyway which is a no-op.